tez-runtime-library/src/main/java/org/apache/tez/dag/library/vertexmanager/VertexManagerWithConcurrentInput.java

```diff
@@ -0,0 +1,234 @@
+    for (Map.Entry<String, EdgeProperty> entry : edges.entrySet()) {
+      if (!CONCURRENT.equals(entry.getValue().getSchedulingType())){
+        throw new TezUncheckedException("All input edges to vertex " + vertexName +
+            "  must be CONCURRENT.");
+      }
+      String srcVertex = entry.getKey();
+      srcVerticesConfigured.put(srcVertex, false);
+      getContext().registerForVertexStateUpdates(srcVertex, EnumSet.of(VertexState.CONFIGURED));
+    }
+
+    try {
+      vertexConfig = TezUtils.createConfFromUserPayload(getContext().getUserPayload());
+    } catch (IOException e) {
+      throw new TezUncheckedException(e);
+    }
+    edgeTriggerType = ConcurrentEdgeTriggerType.valueOf(
+        vertexConfig.get(TezConfiguration.TEZ_CONCURRENT_EDGE_TRIGGER_TYPE,
+            TezConfiguration.TEZ_CONCURRENT_EDGE_TRIGGER_TYPE_DEFAULT));
+    if (!ConcurrentEdgeTriggerType.SOURCE_VERTEX_CONFIGURED.equals(edgeTriggerType)) {
+      // pending TEZ-3999
+      throw new TezUncheckedException("Only support SOURCE_VERTEX_CONFIGURED triggering type for now.");
+    }
+    LOG.info("VertexManagerWithConcurrentInput initialized with edgeTriggerType {}.", edgeTriggerType);
+
+    vertexName = getContext().getVertexName();
+    completedUpstreamTasks = 0;
+  }
+
+  @Override
+  public void onVertexStarted(List<TaskAttemptIdentifier> completions) {
+    onVertexStartedDone.set(true);
+    scheduleTasks();
+  }
+
+  @Override
+  public void onVertexStateUpdated(VertexStateUpdate stateUpdate) {
+    Preconditions.checkArgument(VertexState.CONFIGURED.equals(stateUpdate.getVertexState()),
+        "Received incorrect state notification : " + stateUpdate.getVertexState() + " for vertex: "
+            + stateUpdate.getVertexName() + " in vertex: " + this.vertexName);
+    Preconditions.checkArgument(srcVerticesConfigured.containsKey(stateUpdate.getVertexName()),
+        "Received incorrect vertex notification : " + stateUpdate.getVertexState() + " for vertex: "
+            + stateUpdate.getVertexName() + " in vertex: " + this.vertexName);
+    LOG.info("Received configured notification: " + stateUpdate.getVertexState() + " for vertex: "
+        + stateUpdate.getVertexName() + " in vertex: " + this.vertexName);
+    srcVerticesConfigured.put(stateUpdate.getVertexName(), true);
+    // check for source vertices completely configured
+    boolean checkAllSrcVerticesConfigured = true;
+    for (Map.Entry<String, Boolean> entry : srcVerticesConfigured.entrySet()) {
+      if (!entry.getValue()) {
+        // vertex not configured
+        LOG.info("Waiting for vertex {} in vertex {} ", entry.getKey(), this.vertexName);
+        checkAllSrcVerticesConfigured = false;
+        break;
+      }
+    }
+    allSrcVerticesConfigured = checkAllSrcVerticesConfigured;
+
+    scheduleTasks();
+  }
+
+  @Override
+  public void onSourceTaskCompleted(TaskAttemptIdentifier attempt) {
+    completedUpstreamTasks ++;
```

Comment:
```suggestion
    completedUpstreamTasks++;
```

Reply:
done

Reply:
Maybe log the value of this variable in the log below, something like : completed/total.